mediagrains/gsf.py

```diff
@@ -676,299 +456,50 @@ def get_remaining(self):
-                            only when it is needed. In this case the "skip_data" parameter will be ignored. Default is True.
-        :yields: (Grain, local_id) tuple for each grain
-        :raises GSFDecodeError: If grain is invalid (e.g. no "gbhd" child)
-        """
-        while True:
-            try:
-                with GSFBlock(self.file_data, want_tag="grai") as grai_block:
-                    if grai_block.size == 0:
-                        return  # Terminator block reached
-
-                    local_id = grai_block.read_uint(2)
-
-                    if local_ids is not None and local_id not in local_ids:
-                        continue
-
-                    with GSFBlock(self.file_data, want_tag="gbhd", raise_on_wrong_tag=True) as gbhd_block:
-                        meta = self._decode_gbhd(gbhd_block)
-
-                    data: Optional[bytes] = None
-
-                    if grai_block.has_child_block():
-                        with GSFBlock(self.file_data, want_tag="grdt") as grdt_block:
-                            if grdt_block.get_remaining() > 0:
-                                if load_lazily:
-                                    data = cast(bytes, IOBytes(self.file_data, self.file_data.tell(), grdt_block.get_remaining()))
-                                elif not skip_data:
-                                    data = self.file_data.read(grdt_block.get_remaining())
-
-                yield (self.Grain(meta, data), local_id)
-            except EOFError:
-                return  # We ran out of grains to read and hit EOF
-
-    def grains(self,
-               local_ids: Optional[Sequence[int]] = None,
-               load_lazily: bool = True) -> Iterable[Tuple[GRAIN, int]]:
-        """Generator to get grains from the GSF file. Skips blocks which aren't "grai".
-
-        The file_data will be positioned after the `grai` block.
-
-        :param local_ids: A list of local-ids to include in the output. If None (the default) then all local-ids will be
-                          included
-        :param load_lazily: If True, the grains returned will be designed to lazily load data from the underlying stream
-                            only when it is needed. In this case the "skip_data" parameter will be ignored. Default is True.
-        :yields: (Grain, local_id) tuple for each grain
-        :raises GSFDecodeError: If grain is invalid (e.g. no "gbhd" child)
-        """
-        return self._grains(local_ids=local_ids, skip_data=False, load_lazily=load_lazily)
+class GrainDataLoadingMode (Enum):
+    """This enumeration describes the mode for loading grains from the input.
+
+    For a Non-seekable input:
+        LOAD_IMMEDIATELY -- Grain data will be read as the stream is processed
+        ALWAYS_LOAD_DEFER_IF_POSSIBLE -- Grain data will be read as the stream is processed
+        ALWAYS_DEFER_LOAD_IF_POSSIBLE -- Grain data will be skipped over
```

Comment:
It's unclear to me why this doesn't behave the same as the 2 above. Aren't you forced to read the grain data if the input doesn't support seeking backwards because you would you still want to get the data in the same way as a seekable input? It's also a bit surprising as the default loading mode if it is going to always skip over data.

(I thought seeking forwards on a FIFO/pipe was allowed, but I tried it in C and it returns an error)

Reply:
Have changed